IPIP/0002-denylists-for-ipfs-nodes-and-gateways.md

```diff
@@ -0,0 +1,112 @@
+
+## Motivation
+
+The current [Bad Bit Denylist](https://badbits.dwebops.pub/denylist.json), which is a list of hashed CIDs, has multiple major drawbacks for implementation. One is lacking support for content path, it's hard to block an IPNS path only based on CIDs. The other is lacking description and status code response for each anchor. The description is especially for hashed blocking items, since the hashing is unidirectional, the description can outline the reason why it's been blocked thus increasing maintainability.
+
+A well-thought-out denylist schema can ease the implementation of the denylist, and bring consensus on how content blocking works for gateway and node operators in the community.
+
+## Detailed design
+
+
+### Denylist Schema
+
+Here is the proposed denylist and each field will be explained in detail later.
+
+```js=
+{
+  action: "block",
+  entries: [
+    {
+      type: "cid",
+      content: "bafybeihfqymzmqhbutdd7i4mkq2ltzznzgoshi4r2pnv4hsc2acsojawoe",
+      description: "ipfs quick start",
+      status_code: 410
+    },
+    {
+      type: "content_path",
+      content: "/ipns/example.com",
+      description: "example.com",
+      status_code: 410
+    },
+    {
+      type: "content_path",
+      content: "/ipfs/bafybeihfqymzmqhbutdd7i4mkq2ltzznzgoshi4r2pnv4hsc2acsojawoe",
+      description: "ipfs readme",
+      status_code: 451
+    },
+    {
+      type: "hashed_cid",
+      content: "9056e0f9948c942c16af3564af56d4bb96b6203ad9ccd3425ec628bcd843cc39",
+      description: "sensitive cid that needs to be blocked",
+      status_code: 451
+    },
+    {
+      type: "hashed_content_path",
+      content: "65e60fcaa506ca5b0b49d7aa73df5ba32446bddb4e72a1f8bb5df12eaaaa8745",
+      description: "sensitive content path that needs to be blocked",
+      status_code: 410
+    }
+  ]
+}
+```
+
+#### `action`
+
+Though it's called `denylist`, the only allowed action field here is `block`. Other actions such as `allow` can be added in the future to enable allowlists or other types of content lists.
+
+
+#### Each denylist entry
+
+- `type`: specifies the type of the content should be blocked. E.g. `cid`, `hashed_content_path`.
+- `content`: stores the content that should be blocked according to the type. It's suggested that all CIDv0 needs to be converted into CIDv1 to keep the consistency.
+- `content_path`: the content path needs to be blocked).
+- `description`: description of the CIDs or content paths.
+- `status_code`: status code to be responded for the blocked content. E.g. [410 Gone](https://github.com/ipfs/specs/blob/main/http-gateways/PATH_GATEWAY.md#410-gone); [451 Unavailable For Legal Reasons](https://github.com/ipfs/specs/blob/main/http-gateways/PATH_GATEWAY.md#451-unavailable-for-legal-reasons) or `200 OK` for allowed entry.
```

Comment:
This should be authoritative.
I think anything not `451`, `410` or `200` has no place on a badlist list. I don't want people start to use 3xx or whatever.

Secondly, I don't fully understand why giving out HTTP codes, assuming the goal is to join forces and let gateway operators share ban reasons, **this is really unspecific**,
I would like a machine readable version enum, containing things like `Dos`, `Legal` (maybe with a reason like `Legal Copyright`, `Legal Hatespeech`, `Legal other`, ...), ...
Some gateway operators might prefer faking `500`s or just flat out ignore certain reasons.